src/main/java/org/reaktivity/nukleus/amqp/internal/stream/AmqpServerFactory.java

```diff
@@ -2360,12 +2375,20 @@ private void onDecodeEnd(
                 errorType = end.error().errorList().condition();
             }
             AmqpSession session = sessions.get(decodeChannel);
+            decode:
             if (session != null)
             {
+                session.sessionState = session.sessionState.receivedEnd();
+                if (session.sessionState == AmqpSessionState.ERROR)
+                {
+                    break decode;
+                }
                 session.cleanup(traceId, authorization);
```

Comment:
Suggest handling to `doEncodeEnd` if necessary from inside `session.cleanup(...)` instead.
That will allow the code below, including `doEncodeEnd(...)` and the session state changes to be removed from this method.